### ISSUE TITLE
"Save Draft" / "Saved": focus loss and missing audible feedback

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#15529]
Splitting this out from https://github.com/WordPress/gutenberg/pull/15514#issuecomment-490827361

Overall, the interaction and feedback of the "Save Draft" / "Saved" combo needs some rethinking for better accessibility.

1
Dynamically changing the Save Draft `<button>` to a "Saved" `<span>` produces a focus loss: when using the keyboard and pressing "Save Draft" , the previously focused element gets removed from the DOM and focus is lost.

2
Nothing is announced: when using a screen reader and pressing "Save Draft", there's no audible feedback.

Suggestions:
The "Save Draft" button shouldn't disappear to start with. It should stay in the DOM and still be focused. To avoid this kind of focus losses, in a few other places Gutenberg uses a different pattern: "noop" the button and use `aria-disabled`. For more details, see for example https://github.com/WordPress/gutenberg/issues/15280#issuecomment-490839345

An audible `speak()` message should be used to confirm the draft was saved. Worth noting there are already messages for "Post published." and for "Post reverted to draft." but there's none for "Draft saved".

In all cases, the indication of the "saved" state should be separated from the button:
- buttons should communicate the available action, not the underlying state
- there are other patterns in WordPress to keep the two things separated, see for example the screenshot below:

<img width="305" alt="Screenshot 2019-05-09 at 11 34 59" src="https://user-images.githubusercontent.com/1682452/57445905-ac4b9f80-7253-11e9-96d6-bbfd429d4a61.png">

**Comment by paaljoachim** [comment]
@joedolson Joe

Is this issue still relevant?

**Comment by joedolson** [comment]
Both issues are still present as of Gutenberg 9.7.2. 

It seems like the first issue would be resolved by simply *not* removing the 'Save Draft' button from the DOM while the processing is happening.

Passing the 'Saving' and 'Saved' statuses to speak() sounds like a straightforward patch that just needs to get done; is there something holding this back that I can help with?

**Comment by paaljoachim** [comment]
I will ping various developers and see if we can get some feedback on this issue.
@WordPress/gutenberg-core

**Comment by tellthemachines** [comment]
> It seems like the first issue would be resolved by simply not removing the 'Save Draft' button from the DOM while the processing is happening.

I can't think of any way to stop the button from re-rendering during the process, even if we keep it as a button instead of changing it to a span, because the whole contents of it (label and icon) change. We should instead look into @afercia 's suggestion above:

> To avoid this kind of focus losses, in a few other places Gutenberg uses a different pattern: "noop" the button and use aria-disabled. For more details, see for example #15280 (comment)

Agree that adding a speak() announcement should be straightforward.

I don't think there's anything holding this back apart from nobody's picked it up yet! I won't be able to for the next few weeks, so if anyone's keen - go for it. Otherwise, I'll get back to it when I can.

**Comment by paaljoachim** [comment]
I went ahead and added the "Needs Dev" label.
